subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/tasks/AbstractNativeCompileTask.java

```diff
@@ -65,6 +67,8 @@
     private boolean positionIndependentCode;
     private boolean debug;
     private boolean optimize;
+    private final Property<String> debugInfoFormat;
+    private final Property<String> optimizationLevel;
```

Comment:
I'm a bit hesitant to make those simple `String` which directly represent the compile/linker flag. We should model this a bit more strongly which keeping enough flexibility to allow users to define there own optimization/debug format model. Some model means the same between toolchains like `O1`, `O2`, `Os`, while others are specific to a compiler. Maybe a way to start with this is to identify use case in the sense of what the user do in their project. For example, I see no optimization, some optimization, optimize for speed and optimize for size. At the DSL level, we want to answer what the user wants the how comes at the task/toolchain level.

As for the debugging example, we have embedded symbols (inside the final binary) vs. side-by-side symbols (outside the binary). We also have a variant of the side-by-side symbols for MSVC which allow edit and continue.  For debugging, I may see no debugging symbols, embedded symbols, side-by-side symbols. It may be alright for side-by-side symbols to allow only edit and continue as it is superior to the simple PDB (with a size trade-off that may be negligible at this stage). If I don't mistake, GCC/Clang needs additional steps to make the debugging symbols side-by-side which is something that Gradle would take care of doing properly by adding more tasks or invoke the commands internally to do the work if needed.

Reply:
I'm not saying the model for the debugging format or optimization level should go into this direction but we also have the concept of `CommandLineArgumentProvider` that we could use here.

Reply:
On the simplification side, modeling the optimization as a level enum `NONE` (O0), `COMMON` (O1), `RELEASE` (O2), `AGGRESSIVE` (O3), `SIZE` (Os) could be enough for now. Basically, `O1` is the common, fast, optimization, `O2` is known as maximum optimization without slowing or increasing size, `O3` can be good or not, varies by code (e could leave that out of the model).

Reply:
Not sure I understand all the implications here, but I think you are making a suggestion for a future iteration where we could strongly type these properties allowing for the possibility of these properties getting set not only from within the DSL, but from the command line also.  So maybe for debug, an example would be like
```
gradlew build --debugInfoFormat='/Z7'
```

Reply:
We talked about this a bit last night.  We want to step back from this and gather a bit of information:

- What's meaningful about /Zi vs /Z7 vs other options?  Why would anyone care about these options? We want to avoid adding boolean flags or opaque strings that we pass to the compiler.
- What goes into choosing an optimization level? It makes sense to me that you could have different levels for different components.  We may want to have some modeling for this (like Daniel suggested above).  Do you intend to set an optimization level as a convention for everything (e.g., optimized = O2) or does every project have to make that decision?

A better approach to this might be to look at a way to modify the compiler arguments after/in place of Gradle's translation and before they're sent to the compiler.  That will unblock both issues (debug and optimization level) and avoid some of the deeper modeling questions.

Reply:
Not sure I understand all the implications here, but I think you are making suggestions for a future iteration where the debug/optimization properties are strongly type allowing for the possibility of not only setting from the DSL, but from the command line.  

An example maybe like the following.
```
gradlew build --optimizationLevel=[NONE|COMMON|RELEASE|AGGRESSIVE|SIZE]
```
I may be off base here, but just trying to understand fully.

Reply:
On the /Zi vs /Z7 (VC++ debug symbols) reflects how the debug symbols are outputted. Folks do actually care about this. One puts the symbols out seperate from the artifact (dll, lib etc.) the other embeds them with the output. In our specific project's case, they need /Z7 (embedded) for upstream projects, and not /Zi which is hard coded in the plugin.

Reply:
@lacasseio @big-guy I sort of disagree with the whole idea that command line args should be "modeled". We don't have a similar modeling for Java or any other supported language. What makes native builds any special? The large pool of arguments across multiple compilers and the possible permutation of it doesn't justify the ends. In my opinion, what would go a long way is easy way to add, remove and manipulate these args. They are usually done once and shared in multiple projects, and possibly tweaked in a few. Any modeling will always have exceptions and won't work for some users forcing them to revert back to managing their own, in which case, I would rather just manage all my own rather than having this amalgamation of doing things million different ways.

To make things stranger - One of our use case is to build with full optimization but debug information enabled which is not of the options covered in any of the "modeled" options. The reason for this is to allow having a good performance but still keeping some basic debugging functionality possible. To take it a step further, we have even done where the whole application is compiled with optimizations enabled except for a handful of files where optimization is turned off. These are files where I want to focus my debugging session.

In context of the current change, the intent was not to "design" or "architect" this feature. The intent was to remove the "hard coded" value and allow the scripts to dictate it while at the same moment not breaking existing user functionality. Any design/architect intent should be outside the scope of this specific issue and should be dealt with as a separate independent task.

Reply:
We completely understand the need here. The reason for modeling the argument is to be able to support multiple toolchains without having to bother with the compiler and linker flag for widely known features. It also allows better error message to the user by giving a meaningful message to the user "just trying to get something to work" as well as discoverability of supported features. We understand the need to configure the compiler flag however we want to make a clear distinction between the layers of configurability. At the DSL level, you are configuring "what you are trying to achieve" and at the task level, it should be the "how you are trying to achieve the what". Given that, maybe a better solution without modeling anything would be to switch the `CppCompile` task boolean for debuggable and optimized false, then you can pass whatever argument through the `compileArgs`.

Unfortunately, we can't compare to how Java handle compiler arguments as it's been around for so long that it will be a long cycle before we roll it out. We started experimenting with other plugins like Jacoco.

@fletcher-sumglobal As we move forward, we will to either discouraging users from using plain flags for a strongly modeled feature or allow parsing them from all source (model or plain flags) to keep the values in sync and allows things like tooling  API to acquire the right information. At the moment, we don't have enough modeling to pursue any of those goals.

@cdickson-sum We understand and this is where a model of the debug format could be `embedded` or `side-by-side` or `none`.

Our issue with the previous native modeling (software model) was to give lots of escape hatch. Although those are good for quickly hacking something together, it doesn't help us moving the model forward as well as capturing meaningful use cases. Our decision with the stable native effort is to collect use case first, then write the implementation. In some case, we purposefully don't allow escape hatch because of the complexity of having the right test coverage. As Sterling mentioned above, use case such as what is the use trying to achieve specifically with the optimization? Was there extensive profiling that shows `/O1` or `/O2` for a specific library is the right and only option because of how the compiler works? Is the user want to just say compile this library fast but optimize it a bit? Or is the user want to say I'm releasing the software, so optimize it as much as possible so we can release? Those are all good question allowing us to figure out the right level of modeling at this point.

The goal with Gradle is to make smart decision to help the developer. For example, the use case of compiling the code optimized except for the files you are working with could translate into an automated solution that turns off optimization for all files that are currently modified (compared to the source control). It's a much better solution than allowing the user fiddling with compiler flags per files.

Reply:
> switch the CppCompile task boolean for debuggable and optimized false, then you can pass whatever argument through the `compileArgs`

I might be missing something with your suggestion here. Those are boolean flags and defaulted to false already. The issue is when set false they resolve to /Wi and /O0 correspondingly still. They resolve differently when true.

> The goal with Gradle is to make smart decision to help the developer.

Just don't compromise for the strong flexibility what gradle offers. The options currently available are restrictive, in my humble opinion, and doesn't allow for kind of control I need to support 100+ projects (with all possible permutations of old/new code and multitude of workflows to support).

For our specific use case, we build everything across the board with /W7 with a few exceptions. Why? I have no clue. It was a decision made long before my time. I wouldn't dream of changing it - it will be more like a nightmare. :)

Reply:
I completely understand @amz-shahji and you are right, we are restrictive at the moment on purpose. I forgot about the default when those boolean are set to false. Gradle should allow you to configure everything according to your need. Let look at making [this solution happen](https://github.com/gradle/gradle-native/issues/884#issuecomment-431340483). It's a little variant of your current solution.